core/src/apps/bitcoin/sign_tx/helpers.py

```diff
@@ -339,45 +373,84 @@ def sanitize_tx_meta(tx: PrevTx, coin: CoinInfo) -> PrevTx:
         if tx.branch_id is not None:
             raise wire.DataError("Branch ID not enabled on this coin.")
+
     return tx
 
 
 def sanitize_tx_input(txi: TxInput, coin: CoinInfo) -> TxInput:
     if len(txi.prev_hash) != TX_HASH_SIZE:
         raise wire.DataError("Provided prev_hash is invalid.")
+
     if txi.multisig and txi.script_type not in common.MULTISIG_INPUT_SCRIPT_TYPES:
         raise wire.DataError("Multisig field provided but not expected.")
-    elif not txi.multisig and txi.script_type == InputScriptType.SPENDMULTISIG:
+
+    if not txi.multisig and txi.script_type == InputScriptType.SPENDMULTISIG:
         raise wire.DataError("Multisig details required.")
+
     if txi.address_n and txi.script_type not in common.INTERNAL_INPUT_SCRIPT_TYPES:
         raise wire.DataError("Input's address_n provided but not expected.")
+
+    if coin.decred and txi.script_type not in common.DECRED_INPUT_SCRIPT_TYPES:
+        raise wire.DataError("Invalid script type.")
+
     if not coin.decred and txi.decred_tree is not None:
         raise wire.DataError("Decred details provided but Decred coin not specified.")
+
     if txi.script_type in common.SEGWIT_INPUT_SCRIPT_TYPES or txi.witness is not None:
         if not coin.segwit:
             raise wire.DataError("Segwit not enabled on this coin.")
+
     if txi.commitment_data and not txi.ownership_proof:
         raise wire.DataError("commitment_data field provided but not expected.")
+
     if txi.orig_hash and txi.orig_index is None:
         raise wire.DataError("Missing orig_index field.")
+
     return txi
 
 
 def sanitize_tx_prev_input(txi: PrevInput, coin: CoinInfo) -> PrevInput:
     if len(txi.prev_hash) != TX_HASH_SIZE:
         raise wire.DataError("Provided prev_hash is invalid.")
+
     if not coin.decred and txi.decred_tree is not None:
         raise wire.DataError("Decred details provided but Decred coin not specified.")
+
     return txi
 
 
+def sanitize_decred_commitment_output(txo: TxOutput) -> TxOutput:
+    if txo.script_type != OutputScriptType.PAYTOOPRETURN:
+        raise wire.DataError("Invalid script type.")
+
+    if txo.op_return_data is None:
+        raise wire.DataError("OP_RETURN output without op_return_data")
+
+    if txo.amount != 0:
+        raise wire.DataError("OP_RETURN output with non-zero amount")
+
+    if txo.address or txo.multisig:
+        raise wire.DataError("OP_RETURN output with address or multisig")
+
+    return txo
+
+
 def sanitize_tx_output(txo: TxOutput, coin: CoinInfo) -> TxOutput:
     if txo.multisig and txo.script_type not in common.MULTISIG_OUTPUT_SCRIPT_TYPES:
         raise wire.DataError("Multisig field provided but not expected.")
+
+    if not txo.multisig and txo.script_type == OutputScriptType.PAYTOMULTISIG:
+        raise wire.DataError("Multisig details required.")
+
+    if coin.decred and txo.script_type not in common.DECRED_OUTPUT_SCRIPT_TYPES:
```

Comment:
surprisingly, there's no segwit check on the output side?

we might want to add it instead of the `DECRED_OUTPUT_SCRIPT_TYPES` thing, which should go away

Reply:
https://github.com/trezor/trezor-firmware/pull/1249/commits/7581b3db49ea4d7762df6b63b8f3913d2074e3ba